README.md
## XBraid Quickstart, User Advice, and License 
<!--
  - Copyright (c) 2013, Lawrence Livermore National Security, LLC. 
  - Produced at the Lawrence Livermore National Laboratory. Written by 
  - Jacob Schroder, Rob Falgout, Tzanio Kolev, Ulrike Yang, Veselin 
  - Dobrev, et al. LLNL-CODE-660355. All rights reserved.
  - 
  - This file is part of XBraid. For support, post issues to the XBraid Github page.
  - 
  - This program is free software; you can redistribute it and/or modify it under
  - the terms of the GNU General Public License (as published by the Free Software
  - Foundation) version 2.1 dated February 1999.
  - 
  - This program is distributed in the hope that it will be useful, but WITHOUT ANY
  - WARRANTY; without even the IMPLIED WARRANTY OF MERCHANTABILITY or FITNESS FOR A
  - PARTICULAR PURPOSE. See the terms and conditions of the GNU General Public
  - License for more details.
  - 
  - You should have received a copy of the GNU Lesser General Public License along
  - with this program; if not, write to the Free Software Foundation, Inc., 59
  - Temple Place, Suite 330, Boston, MA 02111-1307 USA
 -->

![](docs/img/logo_with_subtext_2_inch.png)

## What is XBraid? 

XBraid is a parallel-in-time software package.  It implements an
optimal-scaling multigrid solver for the (non)linear systems that arise from
the discretization of problems with evolutionary behavior. 

This code and associated algorithms are developed at [Lawrence Livermore
National Laboratory](https://computation.llnl.gov/projects/parallel-time-integration-multigrid/),
and at collaborating [academic institutions](https://github.com/XBraid/xbraid/wiki/Team), e.g., [UNM](http://www.unm.edu/~jbschroder/).

For our publication list, please go [here](https://github.com/XBraid/xbraid/wiki/Project-Publications). There you will papers on XBraid and various application areas where XBraid has been applied, e.g., fluid dynamics, machine learning, parabolic equations, Burgers' equation, powergrid systems, etc. 

## About XBraid 

Typically, solution algorithms for evolution equations are based on a
time-marching approach, solving sequentially for one time step after the other.
Parallelism in these traditional time-integration techniques is limited to
spatial parallelism.  However, current trends in computer architectures are
leading towards systems with more, but not faster, processors, i.e., clock
speeds are stagnate.  Therefore, faster overall runtimes must come from greater
parallelism. Our approach to achieve such parallelism in time is with multigrid.

In this software, we implement a non-intrusive, optimal-scaling time-parallel
method based on multigrid reduction techniques (multigrid-reduction-in-time or
MGRIT).  A few important points about XBraid are as follows.

- The algorithm enables a scalable parallel-in-time approach by applying multigrid to the time dimension.

- It is designed to be nonintrusive. That is, users apply their existing
  sequential time-stepping code according to our interface, and then XBraid
  does the rest. Users have spent years, sometimes decades, developing the
  right time-stepping scheme for their problem. XBraid allows users to keep
  their schemes, but enjoy parallelism in the time dimension.

- XBraid solves exactly the same problem that the existing sequential
  time-stepping scheme does.

- XBraid is flexible, allowing for a variety of time stepping, relaxation, and
  temporal and spatial coarsening options.

- The full approximation scheme multigrid approach is used to accommodate
  nonlinear problems.

- XBraid written in MPI/C with C++, Fortran 90, and Python interfaces.

- XBraid is released under LGPL 2.1.


## Documentation 

- For examples of using XBraid, see the
  [examples/](https://github.com/XBraid/xbraid/tree/master/examples) and
  [drivers/](https://github.com/XBraid/xbraid/tree/master/drivers) directories,
  and in particular examples/ex-01-*

- See the [release](https://github.com/XBraid/xbraid/releases) page for links
  to precompiled documentation PDFs that go through, step-by-step, how to use
  XBraid.

- For tutorials, see the bottom of our publications 
[page](https://github.com/XBraid/xbraid/wiki/Project-Publications#Tutorials).

- For citing XBraid, see [here](https://github.com/XBraid/xbraid/wiki/Citing-XBraid).


## Advice to Users 

The field of parallel-in-time methods is in many ways under development, and
success has been shown primarily for problems with some parabolic character.
While there are ongoing projects (here and elsewhere) looking at varied
applications such as hyperbolic problems, computational fluid dynamics, power
grids, medical applications, and so on, expectations should take this fact into
account.  That being said, we strongly encourage new users to try our code for
their application.  Every new application has its own issues to address and
this will help us to improve both the algorithm and the software. Please see
our project publications website for our recent
[publications](https://github.com/XBraid/xbraid/wiki/Project-Publications)
concerning some of these varied applications.

For bug reporting, please use the issue tracker here on Github. Please include
as much relevant information as possible, including all the information in the
“VERSION” file located in the bottom most XBraid directory.  For compile and
runtime problems, please also include the machine type, operating system, MPI
implementation, compiler, and any error messages produced. 


## Building XBraid 

-  To specify the compilers, flags and options for your machine, edit
   makefile.inc.  For now, we keep it simple and avoid using configure or
   cmake.

-  To make the library, libbraid.a,
   
         $ make

-  To make the examples
   
         $ make all

-  The makefile lets you pass some parameters like debug with 
   
         $ make debug=yes
   
   or
   
         $ make all debug=yes
   
   It would also be easy to add additional parameters, e.g., to compile with
   insure.  


- To set compilers and library locations, look in makefile.inc
  where you can set up an option for your machine to define simple
  stuff like

       CC = mpicc
       MPICC = mpicc
       MPICXX = mpiCC
       LFLAGS = -lm


## Meaning of the name 

We chose the package name XBraid to stand for Time-Braid, where X is the first
letter in the Greek word for time, Chronos. The algorithm braids together
time-grids of different granularity in order to create a multigrid method and
achieve parallelism in the time dimension.


## License 

This project is released under the LGPL v2.1 license. See files COPYRIGHT and
LICENSE file for full details.

LLNL Release Number: LLNL-CODE-660355



CHANGELOG
----------------------------
Version 3.1.0, 3/2023
----------------------------

This version contains some major new features, including a new capability for
chaotic problems (Delta correction), more detailed timers for performance
analysis, and changes to more easily allow for distributed control variables
during optimization.  Specifically, this release includes

- New Delta correction feature to accelerate MGRIT for nonlinear problems,
    especially chaotic ones.  See user manual and examples/ex-07 for more details,
    including on (i) the reduced rank (storage) option and (ii) deferring Delta
    correction to coarse levels for better efficiency.  This new feature
    builds a basis for the unstable Lyapunov manifold to accurately capture
    the chaotic dynamics.

- Option for user-allocated MPI buffers.  This is important if you need to
    allocate buffers on a device/accelerator (GPU). See braid_SetBufAllocFree().

- Optional user sync function is called at the end (top) of each V-cycle, but
    now with access to the temporal communicator for easier global analyses, e.g.,
    of error estimates over the time dimension.

- Option to solve coarsest grid with relaxation only, which is particularly useful
    for parabolic problems, see braid_SetRelaxOnlyCG().

- Option for more detailed timings.  See braid_SetTimings(), where option 2
    times all user functions and certain important parts of the MGRIT algorithm
    like the coarse-grid solve and MPI Wait.

- Enhancements for easier distributed storage of control variables during while doing
   optimization with MGRIT.
   -> New ability to automatically always store last time point (whether F or C),
      using ulast, see _braid_UGetLast()
   -> New FinalFCRelax feature, that allows for a final FCRelax after Braid
      cycling is finished.  The done flag is True during this FCRelax, and thus
      allows for easy computation of gradient information after Braid cycling.
		If this feature is not used, the user still gets access to the solution
		after Braid halts, but the coarse-grid correction from the final MGRIT
		iteration has not been fully integrated on the finest grid.
   -> New option for easily reverting processor ranks, to allow for backward sweeps
      in time, see braid_SetRevertedRanks()
   -> Adjoint variables are initialized with a non-physical time value of -1.0, to aid
      in programming
   -> braid_defs.h have commented out lines to allow for switch to single precision

- Reduced sensitivity of regression tests to numerical noise.

- Updated documentation to discuss new features, especially Delta correction.


----------------------------
Version 3.0.0, 8/2020
----------------------------

This version contains some major new features, a Python interface, and a new
file structure.

- Added a new periodic feature (activiated through BraidSetPeriodic) that
  allows for solving time-periodic problems efficiently.  The periodic
  structure is replicated on all Braid levels.  See examples/ directory 
  for example codes. 

- Added a new Richardson-based extrapolation option (activated through
  BraidSetRichardsonEstimation).  This option assumes sufficient continuity
  in time and then uses Richardson extrapolation to improve the accuracy 
  order in time on the finest grid.  See examples/ for example codes.

- Added a new Richardson-based error estimation option (activated through
  BraidSetRichardsonEstimation).  This option also assumes sufficent
  continuity in time and then uses Richardson-based techniques to give an 
  error estimate in time on the finest grid.  See examples/ for example codes.

- Added a new Cython (Python) interface.  See braid.pyx and the examples/
  directory for how this works.  In short, Braid can now be called and run
  through Python, including mpi4py, while using other existing Python 
  libraries.

- Added weighted relaxation, which allows for weighting the C-relaxation as
  one would for standard weighted-Jacobi (activated through braid_SetCRelaxWt).
  Experiments show that this weight (when properly chosen) can reduce iterations
  by 10%-30%, for the tested problems.  Common weights are in the range of
  1.0 to 2.0.

- Added an optional new user function called "Sync" that is called only once
  per MPI task at certain points in the code, i.e., at the top of a Braid cycle,
  and after FRefine.  This is different from "Access" which is called for every
  time-point that an MPI task "owns".  Sync is designed to make it easier to
  complete global operations, such as a reduce over the temporal communicator to
  compute some quantity, e.g., a maximum error estimate.

- Split very large files, primarily _braid.c, into many smaller files.  This
  file had become too unwieldy and large.  Also, renamed many files that began
  with "_braid...".  The goal here is to make the filenames easier to grok.

- Added more flexibility for FRefine (which implements FMG/Nested Iteration),
  so that the user can more adaptively control options like the max number of
  levels, the final number of levels, and when to initialize with a sequential
  solution for debugging purposes.

- Updated Make system so that a default make doesn't include MFEM and that an
  error message is printed if hypre isn't found.  Hypre only affects some
  example codes, not Braid itself.

Also,

- Incorporated various bug fixes, including some memory bugs in FRefine

- Updated documentation to discuss new features (weighted-relaxation,
  Richardson-based error estimation and extrapolation, periodic
  functionality, Cython interface, and Sync function.)


----------------------------
Version 2.3.0, 7/11/2018
----------------------------

- New adjoint feature added by Stefanie Guenther, allowing for
  the solution of PDE-constrained optimization problems.  This 
  feature allows XBraid to compute sensitivities (i.e., gradients)
  to be used in an outer optimization loop.  See the new examples 
  examples/ex-01-adjoint.c

_util.c
/*BHEADER**********************************************************************
 * Copyright (c) 2013, Lawrence Livermore National Security, LLC. 
 * Produced at the Lawrence Livermore National Laboratory. Written by 
 * Jacob Schroder, Rob Falgout, Tzanio Kolev, Ulrike Yang, Veselin 
 * Dobrev, et al. LLNL-CODE-660355. All rights reserved.
 * 
 * This file is part of XBraid. For support, post issues to the XBraid Github page.
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License (as published by the Free Software
 * Foundation) version 2.1 dated February 1999.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the IMPLIED WARRANTY OF MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the terms and conditions of the GNU General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc., 59
 * Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 ***********************************************************************EHEADER*/
 

/** \file _util.c
 * \brief Source code for utility routines. See _util.h for more documentation.
 *
 */

#include "_braid.h"
#include "_util.h"

#ifndef DEBUG
#define DEBUG 0
#endif

/*--------------------------------------------------------------------------
 * Project an interval onto a strided index space that contains the index
 * 'index' and has stride 'stride'.  An empty projection is represented by
 * ilower > iupper.
 *--------------------------------------------------------------------------*/
braid_Int
_braid_ProjectInterval( braid_Int   ilower,
                        braid_Int   iupper,
                        braid_Int   index,
                        braid_Int   stride,
                        braid_Int  *pilower,
                        braid_Int  *piupper )
{
   braid_Int  il, iu;

   il = ilower - index;
   iu = iupper - index;

   if ( il <= 0 )
      il = (braid_Int) (il / stride);
   else
      il = (braid_Int) ((il + (stride-1)) / stride);

   if ( iu >= 0 )
      iu = (braid_Int) (iu / stride);
   else
      iu = (braid_Int) ((iu - (stride-1)) / stride);

   *pilower = index + il * stride;
   *piupper = index + iu * stride;

   return _braid_error_flag;
}

braid_Int
_braid_printf( const char *format, ...)
{
   va_list   ap;

   va_start(ap, format);
   if (_braid_printfile != NULL)
   {
      vfprintf(_braid_printfile, format, ap);
      fflush(_braid_printfile);
   }
   else
   {
      vfprintf(stdout, format, ap);
   }
   va_end(ap);

   return _braid_error_flag;
}

braid_Int
_braid_ParFprintfFlush(FILE       *file, 
                       braid_Int   myid,
                       const char *format,
                       ...)
{
   /* Print to file */
   if (myid == 0)
   {
      va_list   ap;
      
      va_start(ap, format);
      vfprintf(file, format, ap);
      va_end(ap);
      fflush(file);
   }

   return _braid_error_flag;
}

braid_Int
_braid_Max(braid_Real  *array, 
           braid_Int    size,
           braid_Real  *max_val)
{  
   braid_Real val;
   braid_Int i;
   if(size > 0)
   {
      val = array[0];
      for(i = 1; i < size; i++)
      {
         if(array[i] > val)
         {  val = array[i]; }
      }
      *max_val = val;
   }
   
   return _braid_error_flag;
}

braid_Int
_braid_GetNEntries(braid_Real   *_array, 
                   braid_Int    array_len, 
                   braid_Int    *k_ptr, 
                   braid_Real   *array)
{   
   braid_Int      start;
   braid_Int      n = *k_ptr;

   if(n < 0)
   {
      /* If negative, copy the last n residual norms */
      n = -n;
      if(array_len < n)
      {
         n = array_len;
      }
      start = array_len - n;
   }
   else
   {
      /* If positive copy the first n residual norms */
      if(array_len < n)
      {
         n = array_len;
      }
      start = 0;
   }

   if(n > 0)
   {
      memcpy(array, &(_array[start]), n*sizeof(braid_Real) );
   }
   else
   {
      array[0] = -1.0;
   }
   (*k_ptr) = n;

   return _braid_error_flag;
}


